Feature/update architecture gamelogic

ibrahimsaad123 opened the pull request.

Commit:
Add Rock entity

Commit:
Refactor MapManager and PlayerManager for enhanced movement and pathfinding integration using BFS.

Commit:
fix piplines and add some test in test_game_logic file for map manager updates

Commit:
Unified resource and effect handling for all card types (Monsters and Buildings), while maintaining distinct categorization and type-specific behaviors.

Commit:
Refactored card handling in PlayerManager: separated card request logic from activation logic for better clarity and flexibility.

AbdulrhmanSayedAli commented on bfs_battle_for_supremacy/game_logic/managers/map_manager.py:
```diff
@@ -1,9 +1,135 @@
+            return
+
+        item = MapManager.current_object
+        for step in path[1:]:
```
this should be delayed in the background (if the steps were done in a for like this the whole app will wait for it and will display the final movement only).

use asyncio.call_later() or a Timer for scheduling each move and run the whole thing in background.

AbdulrhmanSayedAli commented on bfs_battle_for_supremacy/game_logic/entities/monster.py:
```diff
@@ -16,7 +15,6 @@ def __init__(
         consumes,
         **kwargs,
     ):
-        # إزالة `type` من `kwargs` إذا كان موجودًا
```
it's ok you can leave chatgpt comments.

AbdulrhmanSayedAli commented on bfs_battle_for_supremacy/game_logic/managers/cards_manager.py:
```diff
@@ -16,20 +13,14 @@ class CardsManager:
-            card = Building(**card_data)
-        else:
-            card = Card(**card_data)
+        card = Card(**card_data)
```
I think you should delete Monster and Building classes if you chose this approach.

(there is no use of Monster or Building Class in the entire code now except for testing files and this is totally wrong PLEASE REMOVE TESTS IF NOT DONE CORRECTLY).

AbdulrhmanSayedAli commented on bfs_battle_for_supremacy/game_logic/managers/cards_manager.py:
```diff
@@ -76,52 +60,42 @@ def process_recurring_effects(player, enemy_player):
             card.number_of_uses -= 1
 
             if card.valid_for == 0 or card.number_of_uses == 0:
                 expired_cards.append(card)
```
remove expired cards logic (and keep the resources negative as we discussed)

AbdulrhmanSayedAli commented on bfs_battle_for_supremacy/game_logic/managers/cards_manager.py:
```diff
@@ -76,52 +60,42 @@ def process_recurring_effects(player, enemy_player):
             if card.valid_for > 0:
                 card.valid_for -= 1
-
             card.number_of_uses -= 1
```
remove number of uses logic (as we discussed)

AbdulrhmanSayedAli replied:
remove number of uses logic (as we discussed)

AbdulrhmanSayedAli commented on bfs_battle_for_supremacy/game_logic/managers/map_manager.py:
```diff
@@ -1,9 +1,135 @@
+
+    @staticmethod
+    def select_square(square: Square, current_player):
+        if MapManager.selection_counter == 0:
```
the first selection should be always valid (I may select a building to see the details of it or select an enemy's monster to see its stats)

AbdulrhmanSayedAli commented on bfs_battle_for_supremacy/game_logic/managers/map_manager.py:
```diff
@@ -1,9 +1,135 @@
+            else:
+                print("Invalid selection. Square is empty.")
+
+        elif MapManager.selection_counter == 1:
```
here we check if the player previously selected something movable (the player or his monsters) then apply the move.

else it should reselect the current square and remove the old selecting

AbdulrhmanSayedAli commented on bfs_battle_for_supremacy/game_logic/managers/map_manager.py:
```diff
@@ -1,9 +1,135 @@
+
+        else:
+            print("Too many selections. Resetting selection.")
+            MapManager.reset_selection()
```
don't reset selection here (if the player selected some square while his monster is walking this selecting should be totally ignored)

Commit:
Refactor game logic: Enhance MapManager, CardsManager, and PlayerManager functionality

- Updated `MapManager` to improve object selection, movement logic, and target validation.
- Enhanced `CardsManager` to handle card activation, recurring effects, and resource management.
- Improved `PlayerManager` to manage player turns, card usage, and interactions with the game map.
- Streamlined card effects application, monster and building handling, and resource deduction.

Commit:
Comment content test_game_logic file
